calyx-opt/src/passes/static_inliner.rs

```diff
@@ -0,0 +1,388 @@
+            // cond.write_en = %0 ? 1'd1 (since we also have cond.in = %0 ? port)
+            // cond_wire.in = %[1:latency] ? cond.out (since we also have cond_wire.in = %0 ? port)
+            cond["write_en"] = cycle_0_guard ? signal_on["out"];
+            cond_wire["in"] = other_cycles_guard ? cond["out"];
+        );
+        cond_assigns.extend(asgns.to_vec());
+        cond_assigns
+    }
+
+    // inlines the static control `sc` and returns an equivalent single static group
+    fn inline_static_control(
+        sc: &ir::StaticControl,
+        builder: &mut ir::Builder,
+    ) -> ir::RRC<ir::StaticGroup> {
+        match sc {
+            ir::StaticControl::Enable(ir::StaticEnable { group, .. }) => {
+                Rc::clone(group)
+            }
+            ir::StaticControl::Seq(ir::StaticSeq {
+                stmts,
+                latency,
+                attributes,
+            }) => {
+                let seq_group =
+                    builder.add_static_group("static_seq", *latency);
+                let mut seq_group_assigns: Vec<
+                    ir::Assignment<ir::StaticTiming>,
+                > = vec![];
+                let mut cur_offset = 0;
+                for stmt in stmts {
+                    let stmt_latency = stmt.get_latency();
+                    // first recursively call each stmt in seq, and turn each stmt
+                    // into static group g.
+                    let g = StaticInliner::inline_static_control(stmt, builder);
+                    assert!(
+                        g.borrow().get_latency() == stmt_latency,
+                        "static group latency doesn't match static stmt latency"
+                    );
+                    // get the assignments from g
+                    // currently we clone, since we might need these assignments elsewhere
+                    // We could probably do some sort of analysis to see when we need to
+                    // clone vs. can drain
+                    let mut g_assigns: Vec<ir::Assignment<ir::StaticTiming>> =
+                        g.borrow_mut().assignments.clone();
+                    // add cur_offset to each static guard in g_assigns
+                    // and add %[offset, offset + latency] to each assignment in
+                    // g_assigns
+                    StaticInliner::update_assignment_timing(
+                        &mut g_assigns,
+                        cur_offset,
+                        stmt_latency,
+                    );
+                    // add g_assigns to seq_group_assigns
+                    seq_group_assigns.extend(g_assigns.into_iter());
+                    // updates cur_offset so that next stmt gets its static timign
+                    // offset appropriately
+                    cur_offset += stmt_latency;
+                }
+                assert!(
+                    *latency == cur_offset,
+                    "static group latency doesn't match static seq latency"
```

Comment:
I've left a question about this below as well but does the control operator itself need to specify the latency of things for any good reason if the only use is for debugging? It seem it would be more ergonomic if the control operators did not specify latencies of things and relied on the compiler to either infer that latency or simply never reify the latency up till this point where groups are inlined and therefore latency can be computed